### PR TITLE
ref(subscriptions): Include the request body in the result message

### PR DESCRIPTION
**Pull request by tkaemming**
This implements a new result protocol version (2), adding the request body of the query that was run to fetch the original result so that it can be easily replayed later.

### CODE REVIEW

File: snuba/subscriptions/worker.py
Comment (tkaemming) [inline snuba/subscriptions/worker.py:47]: This merges the responsibilities of `SubscriptionExecutor` and `SubscriptionWorker` since they were very closely coupled anyway.

File: snuba/subscriptions/worker.py
Comment (tkaemming) [inline snuba/subscriptions/worker.py:79]: This is the bulk of the previous `SubscriptionExecutor` implementation.

File: snuba/subscriptions/worker.py
Comment (tkaemming) [inline snuba/subscriptions/worker.py:28]: I'd rather the request and result were handled separately, but the request can't be generated before the future execution, otherwise it screws the timings all up because they include request validation, so if there is any queue delay that gets confusingly rolled into the query execution stats.

File: tests/subscriptions/test_codecs.py
Comment (tkaemming) [inline tests/subscriptions/test_codecs.py:79]: This is clunky but it's way easier than generating a `Request` object by hand right now.

File: tests/subscriptions/test_codecs.py
Comment (tkaemming) [inline tests/subscriptions/test_codecs.py:74]: This wasn't actually a valid `Result`. 😅

File: tests/subscriptions/test_worker.py
Comment (tkaemming) [inline tests/subscriptions/test_worker.py:70]: Nothing actually covered the `SubscriptionWorker` before so I thought it made sense to turn the existing executor test into a `SubscriptionWorker` test, which itself became almost an integration test against the dummy Kafka cluster, with the exception that it doesn't use the real `TickConsumer`.

File: tests/subscriptions/test_worker.py
Comment (tkaemming) [inline tests/subscriptions/test_worker.py:106]: It does seem like these responsibilities should be split (constraint enforcement and bucketing.)

File: snuba/subscriptions/worker.py
Comment (tkaemming) [inline snuba/subscriptions/worker.py:134]: This doesn't include the request settings.

File: snuba/cli/subscriptions.py
Comment (tkaemming) [inline snuba/cli/subscriptions.py:20]: This rename made this patch a lot louder than it needed to be, sorry. 😞

File: snuba/subscriptions/worker.py
Comment (fpacifici) [inline snuba/subscriptions/worker.py:79]: I am not sure what the request object in the response is meant to be used for so I am not sure this comment is pertinent. Pay attention, this is not the request you would send to Snuba to run the query. The object you pass to parse_and_run_query is mutable and it gets mutated during the query execution. The object you get from request.body is taken from request.query.body, which is mutated during query execution.
Specifically:
- query extensions are tweaking the query object (the query body in the legacy representation) in place
- dataset query processors are still mutating the query in place in the request.
You do not care about clickhouse query processor since, after #880, we do a copy during translation and the query from that point on is isolated.

In general, as long as the request object contains mutable objects I think it would be safer to return a copy of the request. Of course unless the processed query is what you want to return.

Comment (tkaemming) [reply snuba/subscriptions/worker.py:79]: > I am not sure what the request object in the response is meant to be used for so I am not sure this comment is pertinent.

It's described in the description, but more detail: this is to allow us to double check the query being sent to the database via the subscription consumer is what we expected, and also to allow the same request to be replayed later to ensure that the value hasn't changed. This also would allow the consumer to selectively extract the conditions and to identify the specific data points under aggregation in the subscription (no specific use case for this yet, but we've discussed this being useful in the past.)

I don't think whether this happens before or after query processing doesn't really matter that much to the end consumer — but since there are no guarantees that the processors themselves are idempotent, I agree that it's probably safer for that reason to copy the original request here after it's created to ensure they're not re-applied in a way that might materially change the query.

Ultimately, I don't think this is a great solution for "this is to allow us to double check the query being sent to the database via the subscription consumer is what we expected" but it's the approach with the least entanglement to anything else right now.

File: snuba/subscriptions/worker.py
Comment (fpacifici) [inline snuba/subscriptions/worker.py:140]: Is the consuming code aware of this change already ?

**Review by fpacifici — APPROVED** [review]

File: snuba/subscriptions/worker.py
Comment (tkaemming) [reply snuba/subscriptions/worker.py]: oops, thank

File: snuba/subscriptions/worker.py
Comment (tkaemming) [reply snuba/subscriptions/worker.py:140]: I've talked with @wedamija about it but I will need to wait until he's ready to merge this.

**Pull request merged**
